### DISCUSSION TITLE
Using LVGL and TFT for all XIAO Series | Seeed Studio Wiki

### DISCUSSION THREAD

**Discussion by giscus** [Seeed-Studio/wiki-documents#534] [Q&A]
# Using LVGL and TFT for all XIAO Series | Seeed Studio Wiki

Using LVGL and TFT on the Seeed Studio Round Display for XIAO

https://wiki.seeedstudio.com/using_lvgl_and_tft_on_round_display/

**Comment by emblink** [comment]
It seems that the TFT_eSPI library still does not support the XIAO ESP32-C6 boards as of October 8, 2024. I've purchased 6 XIAO ESP32-C6 boards specifically for display projects, but I cannot use them with this library.

**Reply by bluegizmo83** [reply]
I know this comment is over a year old now, but why would you choose the ESP32-C6 for making a display like this anyway? The C6 is a terrible choice, as it's only a signal core 160mhz ESP32 chip... The C6 is primarily meant to be used for Zigbee/Thread device builds.

**Comment by ikue2000** [comment]
I have just installed SquareLine Studio 1.5.1 and found a couple of issues with above presented code. It does not match the code exported by newer versions of that designer software. Also, the touch panel did not work at all and needed to be fixed.

To get things easier for other starters, please find below the code I needed to modify after I exported by "Create Template Project".

Step 1) Create the interface in SquareLine Studio
Step 2) Export by "Create Template Project"
Step 3) Move "ui.h" and subfolder "src" from the exported data's "libraries" into the "ui" folder where ui.ino is placed
Step 4) Open in Arduino and overwrite the code of "ui.ino" as following:

#include <lvgl.h>
#include <TFT_eSPI.h>

// Changed path
#include <ui.h>

// Added from here
#define USE_TFT_ESPI_LIBRARY
#include "lv_xiao_round_screen.h"
// Added till here


/*Don't forget to set Sketchbook location in File/Preferences to the path of your UI project (the parent foder of this INO file)*/

// Change to your screen resolution
// Removed from here
/*static const uint16_t screenWidth  = 240;
static const uint16_t screenHeight = 240;

static lv_disp_draw_buf_t draw_buf;
static lv_color_t buf[ screenWidth * screenHeight / 10 ];

TFT_eSPI tft = TFT_eSPI(screenWidth, screenHeight); // TFT instance
*/

#if LV_USE_LOG != 0
/* Serial debugging */
void my_print(const char * buf)
{
    Serial.printf(buf);
    Serial.flush();
}
#endif

/* Display flushing */
void my_disp_flush( lv_disp_drv_t *disp, const lv_area_t *area, lv_color_t *color_p )
{
    uint32_t w = ( area->x2 - area->x1 + 1 );
    uint32_t h = ( area->y2 - area->y1 + 1 );

    tft.startWrite();
    tft.setAddrWindow( area->x1, area->y1, w, h );
    tft.pushColors( ( uint16_t * )&color_p->full, w * h, true );
    tft.endWrite();

    lv_disp_flush_ready( disp );
}

/*Read the touchpad*/
void my_touchpad_read( lv_indev_drv_t * indev_driver, lv_indev_data_t * data )
{
    // Changed
    //uint16_t touchX = 0, touchY = 0;
    lv_coord_t touchX, touchY;

    // Changed
    bool touched = chsc6x_is_pressed();//tft.getTouch( &touchX, &touchY, 600 );

    // Added
    if (touched) { chsc6x_get_xy(&touchX, &touchY); }

    if( !touched )
    {
        data->state = LV_INDEV_STATE_REL;
    }
    else
    {
        data->state = LV_INDEV_STATE_PR;

        /*Set the coordinates*/
        data->point.x = touchX;
        data->point.y = touchY;

        Serial.print( "Data x " );
        Serial.println( touchX );

        Serial.print( "Data y " );
        Serial.println( touchY );
    }
}

void setup()
{
    Serial.begin( 115200 ); /* prepare for possible serial debug */

    String LVGL_Arduino = "Hello Arduino! ";
    LVGL_Arduino += String('V') + lv_version_major() + "." + lv_version_minor() + "." + lv_version_patch();

    Serial.println( LVGL_Arduino );
    Serial.println( "I am LVGL_Arduino" );

    lv_init();

    // Added from here
    lv_xiao_disp_init();
    lv_xiao_touch_init();

    // Overwrite callback function as set by lv_xiao_touch_init()
    lv_indev_t * indev = lv_indev_get_next(NULL);
    if (indev) {
        indev->driver->read_cb = my_touchpad_read;
    }
    // Added till here

#if LV_USE_LOG != 0
    lv_log_register_print_cb( my_print ); /* register print function for debugging */
#endif

    // Removed from here
    /*tft.begin();
    tft.setRotation( 3 );

    lv_disp_draw_buf_init( &draw_buf, buf, NULL, screenWidth * screenHeight / 10 );

    //Initialize the display
    static lv_disp_drv_t disp_drv;
    lv_disp_drv_init( &disp_drv );
    // Change the following line to your display resolution
    disp_drv.hor_res = screenWidth;
    disp_drv.ver_res = screenHeight;
    disp_drv.flush_cb = my_disp_flush;
    disp_drv.draw_buf = &draw_buf;
    lv_disp_drv_register( &disp_drv );

    //Initialize the (dummy) input device driver
    static lv_indev_drv_t indev_drv;
    lv_indev_drv_init( &indev_drv );
    indev_drv.type = LV_INDEV_TYPE_POINTER;
    indev_drv.read_cb = my_touchpad_read;
    lv_indev_drv_register( &indev_drv );
    */
    // Removed till here

    ui_init();

    Serial.println( "Setup done" );
}

void loop()
{
    lv_timer_handler(); /* let the GUI do its work */
    delay(5);
}

**Reply by nfs0619** [reply]
Thank you for your valuable feedback and for sharing your solution! We will verify the issue and implement a solution as soon as possible, and update our documentation accordingly.
